# Lab 1.1 - Cleaning and Analyzing Warehouse Data

## Objective
1. Import a CSV dataset using `pandas`
2. Clean the dataset using multiple techniques
3. Answer questions about using the data. 

*Use the challenge dataset to practice cleaning with regex.*

## Rubric

- 6 pts - Contains all required components and uses professional language
- 5 pts - Contains all required components, but uses unprofessional language, formating, etc. 
- 4 pts - Contains some, but not all, of the required components
- 3 pts - Did not submit

## Pandas reference

In addtion to the methods you learned last lab, you may need the following: 

- `df.dtypes` - the data types of the each column
- `df.columns` - the name of each column
- `df.shape` - the number of rows and columns of the dataset
- `df.dropna()` - removes null values
- `pd.todatetime(...)` - sets a given column to a `datetime` object
- `df.column.isunique` - returns True if all data are unique
- `df.column.unique()` - returns all of the uniue values of a column
- `df['col'].replace(...,...)` - replaces values in a column with another value
- `df['col'].value_counts()` - returns the value counts of each unique value
- `df['col'].quantile(0.25) or .quartile(0.75) ` - returns the value at 25% or 75% precent    

## Preparing Data

In this section, we will import and prepare our data to be manipulated

Import `pandas` and `matplotlib.pyplot` here.

In [5]:
import pandas as pd
import matplotlib.pyplot

Use `pandas`' `read_csv()` to create a Dataframe called `warehouse`.

In [6]:
df = pd.read_csv("warehouse_data.csv")

Display the first 5 rows of `warehouse`.

In [7]:
df.head(5)

order_id   customer_id        date nearest_warehouse  \
0  ORD182494  ID6197211592  2019-06-22          Thompson   
1  ORD395518  ID0282825849  2019-12-29          Thompson   
2  ORD494479  ID0579391891  2019-03-02         Nickolson   
3  ORD019224  ID4544561904  2019-01-12         Nickolson   
4  ORD104032  ID6231506320  2019-11-28         Nickolson   

                                       shopping_cart  order_price  \
0  [('Lucent 330S', 1), ('Thunder line', 2), ('iS...        12200   
1        [('Thunder line', 1), ('Universe Note', 2)]         9080   
2               [('Thunder line', 1), ('pearTV', 2)]        10670   
3  [('Universe Note', 1), ('Alcon 10', 2), ('Oliv...        24800   
4  [('Universe Note', 1), ('Olivia x460', 1), ('i...         9145   

   delivery_charges  customer_lat  customer_long  coupon_discount  \
0             79.89    -37.815105     144.932843               10   
1             62.71    -37.802736     144.951118                0   
2             65.87    -37.821302     144.957581               10   
3             57.61    -37.811416     144.973073               15   
4             75.54     37.823859     144.969892               25   

   order_total  season  is_expedited_delivery  distance_to_nearest_warehouse  \
0     11059.89  Winter                   True                         1.2800   
1      9142.71  Summer                  False                         1.1621   
2      9668.87  Autumn                  False                         1.0949   
3     21137.61  Summer                  False                         0.8571   
4      6934.29  Spring                  False                         0.5867   

                              latest_customer_review  is_happy_customer  
0  perfect phone and trusted seller. phone itself...               True  
1  it keeps dropping calls the wifi don't work th...              False  
2            five stars this is a great cheap phone.               True  
3        charger did not fit the charger didn't fit.              False  
4                                    four stars good               True

Display the column names of the Dataframe. 

In [8]:
df.columns

Index(['order_id', 'customer_id', 'date', 'nearest_warehouse', 'shopping_cart',
       'order_price', 'delivery_charges', 'customer_lat', 'customer_long',
       'coupon_discount', 'order_total', 'season', 'is_expedited_delivery',
       'distance_to_nearest_warehouse', 'latest_customer_review',
       'is_happy_customer'],
      dtype='object')

How many data points are in our table?

In [9]:
x, y = df.shape
print(x * y)

8000


We can use one command to answer these three questions: 
- What is the average coupon discount?
- What is the largest order price?
- What is the smallest distance to warehouse?

In [10]:
df.describe()

order_price  delivery_charges  customer_lat  customer_long  \
count     500.000000        500.000000    500.000000     500.000000   
mean    25522.216000         76.658200    -35.835234     144.969494   
std     86333.729169         14.481465     12.045393       0.022720   
min       585.000000         46.350000    -37.827123     144.924967   
25%      7050.000000         65.982500    -37.818222     144.953488   
50%     12807.500000         76.310000    -37.812165     144.965357   
75%     20360.000000         82.555000    -37.805364     144.983985   
max    947691.000000        114.040000     37.826339     145.019837   

       coupon_discount   order_total  distance_to_nearest_warehouse  
count       500.000000  5.000000e+02                     500.000000  
mean         10.890000  3.920967e+04                       2.204224  
std           8.649134  2.741940e+05                       8.812416  
min           0.000000  6.392900e+02                       0.107800  
25%           5.000000  6.454735e+03                       0.751425  
50%          10.000000  1.129396e+04                       1.030100  
75%          15.000000  1.811919e+04                       1.408625  
max          25.000000  5.688270e+06                      94.973400

## Cleaning Data

**Data cleaning** is the process of removing errors and inaccuracies from data to make it ready for analysis In these next steps, we will explore and clean our dataset. 


Check the dataset for null(or empty) values. How many null values are there and in which columns?

In [11]:
print(df.isnull().sum())

order_id                         0
customer_id                      0
date                             0
nearest_warehouse                0
shopping_cart                    0
order_price                      0
delivery_charges                 0
customer_lat                     0
customer_long                    0
coupon_discount                  0
order_total                      0
season                           8
is_expedited_delivery            0
distance_to_nearest_warehouse    0
latest_customer_review           1
is_happy_customer                0
dtype: int64


So `season` and `latest_customer_review` have missing values. Fill the drop the rows with missing values. You should have 491 rows remaining.

In [12]:

df.dropna()

order_id   customer_id        date nearest_warehouse  \
0    ORD182494  ID6197211592  2019-06-22          Thompson   
1    ORD395518  ID0282825849  2019-12-29          Thompson   
2    ORD494479  ID0579391891  2019-03-02         Nickolson   
3    ORD019224  ID4544561904  2019-01-12         Nickolson   
4    ORD104032  ID6231506320  2019-11-28         Nickolson   
..         ...           ...         ...               ...   
495  ORD475510  ID0247417082  2019-08-07         Nickolson   
496  ORD086060  ID0130015098  2019-03-28         Nickolson   
497  ORD079320  ID2707985442  2019-10-16         Nickolson   
498  ORD026546  ID0060088412  2019-05-19         Nickolson   
499  ORD085447  ID5257193650  2019-05-25            Bakers   

                                         shopping_cart  order_price  \
0    [('Lucent 330S', 1), ('Thunder line', 2), ('iS...        12200   
1          [('Thunder line', 1), ('Universe Note', 2)]         9080   
2                 [('Thunder line', 1), ('pearTV', 2)]        10670   
3    [('Universe Note', 1), ('Alcon 10', 2), ('Oliv...        24800   
4    [('Universe Note', 1), ('Olivia x460', 1), ('i...         9145   
..                                                 ...          ...   
495             [('Alcon 10', 2), ('iAssist Line', 2)]       386776   
496  [('pearTV', 2), ('Candle Inferno', 2), ('Toshi...        22120   
497  [('pearTV', 1), ('Alcon 10', 1), ('Lucent 330S...        16490   
498  [('iStream', 1), ('Candle Inferno', 1), ('iAss...         5030   
499  [('iStream', 1), ('Toshika 750', 1), ('Alcon 1...        22370   

     delivery_charges  customer_lat  customer_long  coupon_discount  \
0               79.89    -37.815105     144.932843               10   
1               62.71    -37.802736     144.951118                0   
2               65.87    -37.821302     144.957581               10   
3               57.61    -37.811416     144.973073               15   
4               75.54     37.823859     144.969892               25   
..                ...           ...            ...              ...   
495             59.70    -37.817662     144.970405               10   
496             62.29    -37.815469     144.968083                0   
497            106.98    -37.805979     144.966101                0   
498             79.89     37.820071     144.984435               10   
499             64.62    -37.812231     144.988679               15   

     order_total  season  is_expedited_delivery  \
0       11059.89  Winter                   True   
1        9142.71  Summer                  False   
2        9668.87  Autumn                  False   
3       21137.61  Summer                  False   
4        6934.29  Spring                  False   
..           ...     ...                    ...   
495     20174.70  Winter                   True   
496     22182.29  Autumn                   True   
497     16596.98  Spring                   True   
498      4606.89  Autumn                   True   
499     19079.12  Autumn                   True   

     distance_to_nearest_warehouse  \
0                           1.2800   
1                           1.1621   
2                           1.0949   
3                           0.8571   
4                           0.5867   
..                             ...   
495                         0.1282   
496                         0.3711   
497                         1.4368   
498                         1.3191   
499                         0.6277   

                                latest_customer_review  is_happy_customer  
0    perfect phone and trusted seller. phone itself...               True  
1    it keeps dropping calls the wifi don't work th...              False  
2              five stars this is a great cheap phone.               True  
3          charger did not fit the charger didn't fit.              False  
4                                      four stars good               True  
..                           

Now let's take a look at the data types of each column and see if they make sense.

In [13]:
df.dtypes

order_id                          object
customer_id                       object
date                              object
nearest_warehouse                 object
shopping_cart                     object
order_price                        int64
delivery_charges                 float64
customer_lat                     float64
customer_long                    float64
coupon_discount                    int64
order_total                      float64
season                            object
is_expedited_delivery               bool
distance_to_nearest_warehouse    float64
latest_customer_review            object
is_happy_customer                   bool
dtype: object

Data type of date column is object instead of datetime. So fix this problem and check the datatypes again.

In [14]:
pd.to_datetime(df['date'])

0     2019-06-22
1     2019-12-29
2     2019-03-02
3     2019-01-12
4     2019-11-28
         ...    
495   2019-08-07
496   2019-03-28
497   2019-10-16
498   2019-05-19
499   2019-05-25
Name: date, Length: 500, dtype: datetime64[ns]

In [15]:
df.dtypes

order_id                          object
customer_id                       object
date                              object
nearest_warehouse                 object
shopping_cart                     object
order_price                        int64
delivery_charges                 float64
customer_lat                     float64
customer_long                    float64
coupon_discount                    int64
order_total                      float64
season                            object
is_expedited_delivery               bool
distance_to_nearest_warehouse    float64
latest_customer_review            object
is_happy_customer                   bool
dtype: object

Now let's check for errors. 

First, let's make sure that our data in `order_id` is unique and we don't have any duplicates. 

In [16]:
df['order_id'].is_unique

True

Now, let's check what the unique values of warehouse are.

In [17]:
df['nearest_warehouse'].unique()

array(['Thompson', 'Nickolson', 'Bakers', 'nickolson', 'thompson'],
      dtype=object)

Thompson and thompson should be the same, but here they are being counted as different warehouse. Similarly Nickolson and nickolson are same. Change all of the lowercase names to Title case. 

In [18]:
df['nearest_warehouse'] = df['nearest_warehouse'].replace('thompson', "Thompson")

df['nearest_warehouse'] = df['nearest_warehouse'].replace('nickolson', "Nickolson")


Check the quantities of each warehouse.

In [19]:
df['nearest_warehouse'].value_counts()

nearest_warehouse
Thompson     197
Nickolson    184
Bakers       119
Name: count, dtype: int64

The value counts should be Thompson: 197, Nickolson: 184, Bakers: 119

Check the counts and unique values of `is_happy_customer`.

In [20]:
df['is_happy_customer'].value_counts()

is_happy_customer
True     359
False    141
Name: count, dtype: int64

Create a table with `order_price`, `delivery_charges`, `coupon_discount`, and `order_total`. Look at the first 10 rows. Is the `order_total` column correct?

In [21]:
df.head()
del df['customer_id']
del df['order_id'] 
del df['date']
#del df['order_id']
df.head()

nearest_warehouse                                      shopping_cart  \
0          Thompson  [('Lucent 330S', 1), ('Thunder line', 2), ('iS...   
1          Thompson        [('Thunder line', 1), ('Universe Note', 2)]   
2         Nickolson               [('Thunder line', 1), ('pearTV', 2)]   
3         Nickolson  [('Universe Note', 1), ('Alcon 10', 2), ('Oliv...   
4         Nickolson  [('Universe Note', 1), ('Olivia x460', 1), ('i...   

   order_price  delivery_charges  customer_lat  customer_long  \
0        12200             79.89    -37.815105     144.932843   
1         9080             62.71    -37.802736     144.951118   
2        10670             65.87    -37.821302     144.957581   
3        24800             57.61    -37.811416     144.973073   
4         9145             75.54     37.823859     144.969892   

   coupon_discount  order_total  season  is_expedited_delivery  \
0               10     11059.89  Winter                   True   
1                0      9142.71  Summer                  False   
2               10      9668.87  Autumn                  False   
3               15     21137.61  Summer                  False   
4               25      6934.29  Spring                  False   

   distance_to_nearest_warehouse  \
0                         1.2800   
1                         1.1621   
2                         1.0949   
3                         0.8571   
4                         0.5867   

                              latest_customer_review  is_happy_customer  
0  perfect phone and trusted seller. phone itself...               True  
1  it keeps dropping calls the wifi don't work th...              False  
2            five stars this is a great cheap phone.               True  
3        charger did not fit the charger didn't fit.              False  
4                                    four stars good               True

### Remove Outliers

We are going to remove statistical outliers in `order_total`. To remove outliers, follow this process:

- Find upper and lower quartiles
- Find IQR range
- Find upper and lower limit
- Remove elements 

In [ ]:
#print(order_total)
df['order_total'].quantile([0.25, 1.0])
Q1 = df['order_total'].quantile(0.25)
Q3 = df['order_total'].quantile(0.75)
IQR = Q3 - Q1

0.25       6454.735
1.00    5688269.600
Name: order_total, dtype: float64

In [25]:
Q1 = df['order_total'].quantile(0.25)
Q3 = df['order_total'].quantile(0.75)
IQR = Q3 - Q1
print(IQR)

11664.4525


In [26]:
print(Q1, Q3)

6454.735 18119.1875


In [ ]:
Q1 = df['order_total'].quantile(0.25)
Q3 = df['order_total'].quantile(0.75)
IQR = Q3 - Q1

There are 18 records with outliers to remove. 

## Explore Data

Now that our data is clean, we can answer questions about it!

What is the largest order?

What is the smallest order?

Without outliers, what are the top 5 most expensive orders?

What warehouse is the closest to most customers? Create a visualization.

What percentage of customers are satisfied with their purchase? Create a visualization.